api/v1/helpers/verbs/heartbeatUtils.js

```diff
@@ -35,76 +30,106 @@ const NOCHANGE = 4;
+      if (inOld && !inNew) change = DELETED;
+      if (!inOld && inNew) change = ADDED;
+      if (inOld && inNew) change = UPDATED;
 
-    //determine the change type
-    let change;
-    const inOld = oldCollectors.includes(collectorName);
-    const inNew = newCollectors.includes(collectorName);
-    if (inOld && !inNew) change = DELETED;
-    if (!inOld && inNew) change = ADDED;
-    if (inOld && inNew) change = UPDATED;
+      // determine if this generator has already been changed since the last heartbeat
+      const alreadyAdded = changes.added.includes(genId);
+      const alreadyDeleted = changes.deleted.includes(genId);
+      const alreadyUpdated = changes.updated.includes(genId);
+      const alreadyChanged = alreadyAdded || alreadyDeleted || alreadyUpdated;
 
-    //determine if this generator has already been changed since the last heartbeat
-    const addedGenerators = collectorMap[collectorName].added;
-    const deletedGenerators = collectorMap[collectorName].deleted;
-    const updatedGenerators = collectorMap[collectorName].updated;
-    const alreadyAdded = addedGenerators.find((g) => g.id === generator.id);
-    const alreadyDeleted = deletedGenerators.find((g) => g.id === generator.id);
-    const alreadyUpdated = updatedGenerators.find((g) => g.id === generator.id);
-    const alreadyChanged = alreadyAdded || alreadyDeleted || alreadyUpdated;
-
-    //reset the change map to account for previous changes
-    if (alreadyChanged) {
-      if (alreadyAdded && change === DELETED) {
-        removeFromList(addedGenerators);
-        change = NOCHANGE;
-      } else if (alreadyDeleted && change === ADDED) {
-        removeFromList(deletedGenerators);
-        change = UPDATED;
-      } else if (alreadyUpdated && change === DELETED) {
-        removeFromList(updatedGenerators);
-        change = DELETED;
-      } else {
-        change = NOCHANGE;
+      // account for previous changes
+      let undoChange;
+      if (alreadyChanged) {
+        if (alreadyAdded && change === DELETED) {
+          undoChange = ADDED;
+          change = NOCHANGE;
+        } else if (alreadyDeleted && change === ADDED) {
+          undoChange = DELETED;
+          change = UPDATED;
+        } else if (alreadyUpdated && change === DELETED) {
+          undoChange = UPDATED;
+          change = DELETED;
+        } else {
+          change = NOCHANGE;
+        }
       }
-    }
 
-    //update collectorMap based on the change type
-    if (change === ADDED) {
-      addedGenerators.push(generator);
-    } else if (change === DELETED) {
-      deletedGenerators.push(generator);
-    } else if (change === UPDATED) {
-      updatedGenerators.push(generator);
-    }
+      // update redis
+      promises.push(removeFromSet(collectorName, undoChange, genId));
+      promises.push(addToSet(collectorName, change, genId));
+    });
 
+    return Promise.all(promises);
   });
+} // trackGeneratorChanges
+
+function getChanges(collectorName) {
```

Comment:
Can we please add comments to the functions that do not have them and declare them before they are used?